tests/Common/ConfigurationMetadataTest.php

```diff
@@ -0,0 +1,405 @@
+        $this->validateMetadataEquality($updatedMetadata[0], $listConfigurationMetadata[1]);
+
+        // add new metadata to the default branch
+        $moreMetadata = [
+            [
+                'key' => 'KBC.addMetadata.thirdMetadata',
+                'value' => 'my awesome next value',
+            ]
+        ];
+
+        $configurationMetadataOptions = (new ConfigurationMetadata($transformationMain1Options))
+            ->setMetadata($moreMetadata);
+        $components->addConfigurationMetadata($configurationMetadataOptions);
+
+        // update development branch config metadata
+        $configurationMetadataOptions = (new ConfigurationMetadata($transformationMain1Options))
+            ->setMetadata([
+                [
+                    'key' => 'someMetadataKey',
+                    'value' => 'development branch value',
+                ]
+            ]);
+        $branchComponents->addConfigurationMetadata($configurationMetadataOptions);
+
+        $branchComponents->resetToDefault('transformation', 'main-1');
+
+        $listConfigurationMetadata = $branchComponents->listConfigurationMetadata((new ListConfigurationMetadataOptions())
+            ->setComponentId('transformation')
+            ->setConfigurationId('main-1'));
+        self::assertCount(3, $listConfigurationMetadata);
+
+        $this->validateMetadataEquality(self::TEST_METADATA[0], $listConfigurationMetadata[0]);
+        $this->validateMetadataEquality($updatedMetadata[0], $listConfigurationMetadata[1]);
+        $this->validateMetadataEquality($moreMetadata[0], $listConfigurationMetadata[2]);
+    }
+
+    public function testConfigMetadataRestrictionsForReadOnlyUser()
+    {
+        $guestClient = $this->getClientForToken(STORAGE_API_READ_ONLY_TOKEN);
+
+        // create new configurations in main branch
+        $components = new Components($this->_client);
+        $transformationMain1Options = (new Configuration())
+            ->setComponentId('transformation')
+            ->setConfigurationId('main-1')
+            ->setName('Main 1')
+            ->setConfiguration(['test' => 'false']);
+
+        $components->addConfiguration($transformationMain1Options);
+        $configurationMetadataOptions = (new ConfigurationMetadata($transformationMain1Options))
+            ->setMetadata(self::TEST_METADATA);
+
+        try {
+            $components = new Components($guestClient);
+            $components->addConfigurationMetadata($configurationMetadataOptions);
+            $this->fail('should fail, insufficiently permission');
+        } catch (ClientException $e) {
+            $this->assertContains('Configuration manipulation is restricted for your user role ', $e->getMessage());
+            $this->assertSame(403, $e->getCode());
+        }
+    }
+
+    private function validateMetadataEquality(array $expected, array $actual)
+    {
+        foreach ($expected as $key => $value) {
+            self::assertArrayHasKey($key, $actual);
```

Comment:
používáme $this-> ne self::

Reply:
Fair enough, koukám, že je to na pár místech použité takto... akorát jsou to prakticky jen Exasol, Teradata a něco okolo synapse :D